[inductor] Memory planning

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102625

This implements static memory planning. It's disabled by default while we examine its performance.

We use a greedy-by-size approach. For dynamic shapes, the sizes of the example inputs are used as estimates when making planning decisions. We generate expressions to calculate the actual memory offsets and sizes at runtime when the values of the dynamic shapes are known. In order to simplify these calculations, we have organized the allocations into a tree that branches on space (address offsets) and time (live ranges). Finally, we need to align these offsets, so we have added an `align` sympy Expr to express these calculations.

Some limitations:
1. It is only enabled during inference for now. Enabling it for training increases peak memory usage as we allocate all the memory needed for training upfront, before freeing the memory allocated during inference. We can probably address this by doing planning for both the inference and training passes together.
2. It doesn't work with PyTorch Distributed, because kernels like `AllGatherIntoTensor` codegen strings which do memory operations. We can fix this down the line by having them emit MemoryPlanningLines instead.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

jansel committed:
[inductor] Memory planning

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 anijain2305

[ghstack-poisoned]

ipiszy commented on torch/_inductor/codegen/memory_planning.py:
```diff
@@ -0,0 +1,798 @@
+    """
+
+    wrapper: Any
+    pools: AllocationPools = dataclasses.field(default_factory=AllocationPools)
```
Please add comments for pools and buffer_groups.

ipiszy commented on torch/_inductor/codegen/memory_planning.py:
```diff
@@ -0,0 +1,798 @@
+    codegen.
+    """
+
+    wrapper: Any
```
Where is wrapper assigned? What does it represent?

jansel replied:
It is the wrapper codegen object in wrapper.py

ipiszy commented on torch/_inductor/codegen/memory_planning.py:
```diff
@@ -0,0 +1,798 @@
+            s for s in self.allocations if s.get_live_ranges().overlaps(block_live)
+        ]
+        if len(overlapping) > 1:
+            # TODO(jansel): we could try harder here by merging overlapping in space
```
It seems that it's still possible to allocate the block if there is enough remaining space?

ipiszy replied:
FYI this is the memory planner in AIT: https://github.com/facebookincubator/AITemplate/blob/main/python/aitemplate/compiler/transform/memory_planning.py#L196, which tries to fill in gaps between different allocations.

jansel replied:
This is using roughly the same algorithm as the AIT one.  They both do greedy by size.

The difficulty here, and the reason for the tree, is dealing with dynamic shapes.  Which the original paper on this doesn't support.  If you have one tensor of size `s0` and another of size `s1` you don't nessarily know which one of those is bigger.  The tree allows us to compute symbolic boundaries (e.g. `align(max(s0, s1))`) that will turn into runtime dynamic shape expressions.

This specific line of code isn't actually hit on the models I tested, so doesn't matter.

ipiszy commented on torch/_inductor/codegen/memory_planning.py:
```diff
@@ -0,0 +1,798 @@
+
+
+@dataclasses.dataclass
+class MemoryPlanner(object):
```
It would be really helpful to describe the algorithm of memory planning somewhere in this file for better readability.

ipiszy reviewed (commented):
Thanks @jansel ! Left some comments. As a new developer, I feel it would be very beneficial if there are more comments in code, thanks!

jansel commented on torch/_inductor/codegen/memory_planning.py:
```diff
@@ -0,0 +1,798 @@
+                name = line.node.get_name()
+                assert name not in name_to_group
+                name_to_group[name] = BufferGroup(line.node)
+            elif isinstance(line, ReuseLine):
```
No, it is used to represent in-place reuse.  If an input to a pointwise kernel is no longer needed, we might write outputs to the same memory as inputs.

jansel commented on torch/_inductor/codegen/memory_planning.py:
```diff
@@ -0,0 +1,798 @@
+        """Number of bytes needed at runtime"""
+        raise NotImplementedError()
+
+    def finalize(self, pool, offset) -> "AllocationTreeNode":
```
> Does it represent the offset of this Allocation in the AllocationPool?
yes

jansel committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 anijain2305

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 anijain2305

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 anijain2305

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 anijain2305

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

swolchok commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1575,10 +1631,47 @@ def codegen_int_array_var(self, int_array: str, writer=None):
-        size = self.codegen_shape_tuple(tuple(buffer.get_size()))
-        stride = self.codegen_shape_tuple(tuple(buffer.get_stride()))
+        # outputs are passed-in in the AOT mode
+        if self.use_preallocated_output(buffer):
```
can confirm, this function doesn't seem to exist

int3 replied:
yeah this was an artifact of rebasing. this PR isn't intended for review yet

int3 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -558,6 +555,32 @@ def generate(self):
         return result.getvaluewithlinemap()
 
+    def memory_plan(self):
+        if config.memory_planning and not V.graph.cpp_wrapper:
```
yeah I guess so. I'll look into enabling it

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

amitaga commented:
@amitaga has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50069118).

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Memory planning"


This implements static memory planning. It's disabled by default while we examine its performance.

We use a greedy-by-size approach. For dynamic shapes, the sizes of the example inputs are used as estimates when making planning decisions. We generate expressions to calculate the actual memory offsets and sizes at runtime when the values of the dynamic shapes are known. In order to simplify these calculations, we have organized the allocations into a tree that branches on space (address offsets) and time (live ranges). Finally, we need to align these offsets, so we have added an `align` sympy Expr to express these calculations.

Some limitations:
1. It is only enabled during inference for now. Enabling it for training increases peak memory usage as we allocate all the memory needed for training upfront, before freeing the memory allocated during inference. We can probably address this by doing planning for both the inference and training passes together.
2. It doesn't work with PyTorch Distributed, because kernels like `AllGatherIntoTensor` codegen strings which do memory operations. We can fix this down the line by having them emit MemoryPlanningLines instead.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

int3 commented:
Hm, it seems like I can't add @jansel as a reviewer. Should I create a new PR?

int3 commented:
Created https://github.com/pytorch/pytorch/pull/112178
